File: zettel/request.py
from .filenames import fix_path

def delete_note_keywords():
    return "DELETE FROM keywords WHERE note = :note"

def delete_note_links():
    return "DELETE FROM links WHERE src = :src"

def delete_note_folgezettels():
    return "DELETE FROM folgezettels WHERE outline = :outline"

def add_note():
    return """
        INSERT INTO notes (filename, title) VALUES (:filename, :title)
            ON CONFLICT (filename) DO UPDATE SET title = :title
    """

def add_link():
    return """
        INSERT INTO links (src, dest, description)
            VALUES (:src, :dest, :description)
                ON CONFLICT (src, dest) DO UPDATE
                    SET description = description
    """

def add_keyword():
    return """
        INSERT OR IGNORE INTO keywords (note, keyword)
            VALUES (:note, :keyword)
    """

def add_folgezettel():
    return """
        INSERT OR IGNORE INTO folgezettels (outline, note, seqnum)
            VALUES (:outline, :note, :seqnum)
    """

def transform_link_params(params):
    src = params.get("src")
    dest = fix_path(params.get("dest"), src)
    if not dest:
        return None
    return {
        "src": src,
        "dest": dest,
        "description": params.get("description"),
    }

def to_sql(req):
    t = req.get("type", "")
    if t == "note":
        return note(req)
    elif t == "link":
        return link(req)
    elif t == "keyword":
        return keyword(req)
    return ("",)

File: zettel/request.lua
-- construct request data for zettel sqlite3

local M = {}

local function json_string(s)
    return string.format('"%s"', s:gsub('"', [[\"]]))
end

function M.note(title, filename)
    local tmpl = [[
    {
        "type": "note",
        "data": {
            "title": @title@,
            "filename": @filename@
        }
    }
    ]]
    return tmpl:gsub("@filename@", json_string(filename))
        :gsub("@title@", json_string(title))
end

function M.link(src, dest, description)
    local tmpl = [[
    {
        "type": "link",
        "data": {
            "src": @src@,
            "dest": @dest@,
            "description": @description@
        }
    }
    ]]
    return tmpl:gsub("@description@", json_string(description))
        :gsub("@dest@", json_string(dest))
        :gsub("@src@", json_string(src))
end

function M.keyword(note, keyword)
    -- note (filename) and keyword (tag)
    local tmpl = [[
    {
        "type": "keyword",
        "data": {
            "note": @note@,
            "keyword": @keyword@
        }
    }
    ]]
    return tmpl:gsub("@keyword@", json_string(keyword))
        :gsub("@note@", json_string(note))
end

function M.sequence(outline, prev_note, next_note)
    local tmpl = [[
    {
        "type": "sequence",
        "data": {
            "prev": @prev@,
            "next": @next@,
            "outline": @outline@
        }
    }
    ]]
    return tmpl:gsub("@outline@", json_string(outline))
        :gsub("@next@", json_string(next_note))
        :gsub("@prev@", json_string(prev_note))
end

function M.folgezettel(outline, note, seqnum)
    local tmpl = [[
    {
        "type": "folgezettel",
        "data": {
            "outline": @outline@,
            "note": @note@,
            "seqnum": @seqnum@
        }
    }
    ]]
    return tmpl:gsub("@seqnum@", json_string(seqnum))
        :gsub("@note@", json_string(note))
        :gsub("@outline@", json_string(outline))
end

return M
